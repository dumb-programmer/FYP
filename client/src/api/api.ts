export function login(data: { email: string; password: string }) {
  return fetch("http://localhost:3000/login", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
    credentials: "include",
  });
}

export function signup(data: {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
}) {
  return fetch("http://localhost:3000/signup", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
  });
}

export function getMessages(chatId: string, page: number) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function searchMessages(chatId: string, query: string, page: number) {
  return fetch(
    `http://localhost:3000/chats/${chatId}/messages?query=${query}&page=${page}`,
    {
      mode: "cors",
      credentials: "include",
    }
  );
}

export function deleteMessage(chatId: string, messageId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages/${messageId}`, {
    method: "DELETE",
    mode: "cors",
    credentials: "include",
  });
}

export function sendPrompt(data: { prompt: string }, chatId?: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "POST",
    mode: "cors",
    credentials: "include",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
  });
}

export function getChats(page: number) {
  return fetch(`http://localhost:3000/chats?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function updateChat(chatId: string, name: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "PATCH",
    body: JSON.stringify({ name }),
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
    mode: "cors",
  });
}

export function deleteChat(chatId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "DELETE",
    credentials: "include",
    mode: "cors",
  });
}
