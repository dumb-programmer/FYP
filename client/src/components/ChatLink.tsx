import { useEffect, useRef, useState } from "react";
import { NavLink } from "react-router-dom";
import EditChatForm from "./EditChatForm";
import { EllipsisHorizontalIcon, PencilIcon, TrashIcon } from "@heroicons/react/16/solid";
import ConfirmationModal from "./ConfirmationModal";
import { deleteChat } from "@/api/api";

export default function ChatLink({ chat, refetch }) {
    const [edit, setEdit] = useState(false);
    const [dropdownOpen, setDropdownOpen] = useState(false);
    const deleteChatModalRef = useRef<HTMLDialogElement>(null);

    useEffect(() => {
        const onClick = () => {
            if (dropdownOpen) {
                setDropdownOpen(false);
            }
        };
        document.addEventListener("click", onClick);


        return () => {
            document.removeEventListener("click", onClick);
        }

    }, [dropdownOpen]);

    return <div>
        {
            !edit ? <div className="relative">
                <li>
                    <NavLink className="flex justify-between" to={`/${chat._id}`}>
                        {chat.name}
                        <button className="btn btn-ghost btn-xs" onClick={(e) => {
                            e.preventDefault()
                            e.stopPropagation()
                            setDropdownOpen(open => !open)
                        }}>
                            <EllipsisHorizontalIcon height={20} width={20} className="text-fuchsia-50" />
                        </button>
                    </NavLink>
                </li>
                <div className={`absolute right-0 dropdown${dropdownOpen ? "dropdown-open" : ""}`}>
                    <ul className="dropdown-content z-[1] menu p-2 shadow bg-base-100 rounded-box w-52">
                        <li>
                            <button onClick={() => setEdit(true)}>
                                <PencilIcon height={20} width={20} />
                                Rename
                            </button>
                        </li>
                        <li>
                            <button onClick={() => deleteChatModalRef.current?.showModal()}>
                                <TrashIcon height={20} width={20} />
                                Delete
                            </button>
                        </li>
                    </ul>
                </div>
                <div className="flex gap-2">
                    <ConfirmationModal dialogRef={deleteChatModalRef} onDelete={async () => {
                        const response = await deleteChat(chat._id);
                        if (response.ok) {
                            refetch()
                        }
                    }} />
                </div>
            </div> : <EditChatForm chat={chat} refetch={refetch} onCancel={() => setEdit(false)} />
        }
    </div>
}