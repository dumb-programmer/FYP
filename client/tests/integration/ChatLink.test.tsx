import React from "react";
import { render, screen } from "@testing-library/react";
import { expect, describe, it, vi, beforeEach, afterEach } from "vitest";
import ChatLink from "../../src/components/ChatLink";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router-dom";
import { deleteChat } from "../../src/api/api";

vi.mock("../../src/api/api.ts", async () => ({ ...await vi.importActual("../../src/api/api"), deleteChat: vi.fn().mockReturnValue({ ok: true }) }))

const chatId = "123";
const chat = { _id: chatId, name: "chat" };
const mockRefetch = vi.fn();
const renderComponent = () => render(<MemoryRouter><ChatLink chat={chat} refetch={mockRefetch} /></MemoryRouter>);

describe("ChatLink", () => {
    beforeEach(() => {
        HTMLDialogElement.prototype.showModal = vi.fn(function () {
            this.style.display = 'block';
            this.setAttribute('open', '');
        });

        HTMLDialogElement.prototype.close = vi.fn(function () {
            this.style.display = 'none';
            this.removeAttribute('open');
        });
    });

    afterEach(() => {
        vi.clearAllMocks()
    })

    it("renders correctly", () => {
        const { container } = renderComponent();

        expect(container).toMatchSnapshot();
    });

    it("clicking on dropdown button opens the dropdown with rename and edit buttons", async () => {
        renderComponent();

        const dropdown = screen.getByTestId("chat-actions-dropdown")

        await userEvent.click(dropdown);

        expect(screen.getByRole("button", { name: /rename/i })).toBeInTheDocument();
        expect(screen.getByRole("button", { name: /delete/i })).toBeInTheDocument();
    });

    it("rename button opens the EditChatForm", async () => {
        renderComponent();

        const dropdown = screen.getByTestId("chat-actions-dropdown")

        await userEvent.click(dropdown);

        await userEvent.click(screen.getByRole("button", { name: /rename/i }));

        const input = screen.getByRole("textbox");
        const saveBtn = screen.getByRole("button", { name: /submit/i });
        const cancelBtn = screen.getByRole("button", { name: /cancel/i });

        expect(input).toBeInTheDocument();
        expect(input.value).toEqual(chat.name);
        expect(saveBtn).toBeInTheDocument();
        expect(cancelBtn).toBeInTheDocument();
    });

    it("delete button opens the DeleteConfirmationModal", async () => {
        renderComponent();

        const dropdown = screen.getByTestId("chat-actions-dropdown")

        await userEvent.click(dropdown);

        await userEvent.click(screen.getByRole("button", { name: /delete/i }));

        expect(screen.getByRole("dialog")).toBeInTheDocument();
    });

    it("delete button onDelete calls the deleteChat function and refetches the chats", async () => {
        renderComponent();

        const dropdown = screen.getByTestId("chat-actions-dropdown")

        await userEvent.click(dropdown);

        await userEvent.click(screen.getByRole("button", { name: /delete/i }));

        expect(screen.getByRole("dialog")).toBeInTheDocument();

        await userEvent.click(screen.getByTestId("delete-confirmation-dialog-delete-btn"))

        expect(deleteChat).toHaveBeenCalled();
        expect(deleteChat).toHaveBeenCalledWith(chat._id);
        expect(mockRefetch).toHaveBeenCalled();
    });

    it("delete button onCancel closes the DeleteConfirmationModal", async () => {
        renderComponent();

        const dropdown = screen.getByTestId("chat-actions-dropdown")

        await userEvent.click(dropdown);

        await userEvent.click(screen.getByRole("button", { name: /delete/i }));

        expect(screen.getByRole("dialog")).toBeVisible();

        await userEvent.click(screen.getByTestId("delete-confirmation-dialog-cancel-btn"));

        expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
    });
})