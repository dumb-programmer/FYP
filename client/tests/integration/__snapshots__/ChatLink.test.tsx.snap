// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChatLink > renders correctly 1`] = `
<div>
  <div
    class="relative"
  >
    <li
      class="-z-0"
    >
      <a
        class="flex justify-between"
        href="/123"
      >
        chat
        <button
          class="btn btn-ghost btn-xs"
        >
          <svg
            aria-hidden="true"
            class="text-fuchsia-50"
            data-slot="icon"
            fill="currentColor"
            height="20"
            viewBox="0 0 16 16"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2 8a1.5 1.5 0 1 1 3 0 1.5 1.5 0 0 1-3 0ZM6.5 8a1.5 1.5 0 1 1 3 0 1.5 1.5 0 0 1-3 0ZM12.5 6.5a1.5 1.5 0 1 0 0 3 1.5 1.5 0 0 0 0-3Z"
            />
          </svg>
        </button>
      </a>
    </li>
    <div
      class="absolute right-0 z-10 overflow-hidden dropdown"
      data-testid="chat-actions-dropdown"
    >
      <ul
        class="dropdown-content menu p-2 shadow bg-base-100 rounded-box w-52"
      >
        <li>
          <button
            class="text-xs"
          >
            <svg
              aria-hidden="true"
              data-slot="icon"
              fill="currentColor"
              height="15"
              viewBox="0 0 16 16"
              width="15"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M11.013 2.513a1.75 1.75 0 0 1 2.475 2.474L6.226 12.25a2.751 2.751 0 0 1-.892.596l-2.047.848a.75.75 0 0 1-.98-.98l.848-2.047a2.75 2.75 0 0 1 .596-.892l7.262-7.261Z"
                fill-rule="evenodd"
              />
            </svg>
            Rename
          </button>
        </li>
        <li>
          <button
            class="text-xs"
          >
            <svg
              aria-hidden="true"
              data-slot="icon"
              fill="currentColor"
              height="15"
              viewBox="0 0 16 16"
              width="15"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M5 3.25V4H2.75a.75.75 0 0 0 0 1.5h.3l.815 8.15A1.5 1.5 0 0 0 5.357 15h5.285a1.5 1.5 0 0 0 1.493-1.35l.815-8.15h.3a.75.75 0 0 0 0-1.5H11v-.75A2.25 2.25 0 0 0 8.75 1h-1.5A2.25 2.25 0 0 0 5 3.25Zm2.25-.75a.75.75 0 0 0-.75.75V4h3v-.75a.75.75 0 0 0-.75-.75h-1.5ZM6.05 6a.75.75 0 0 1 .787.713l.275 5.5a.75.75 0 0 1-1.498.075l-.275-5.5A.75.75 0 0 1 6.05 6Zm3.9 0a.75.75 0 0 1 .712.787l-.275 5.5a.75.75 0 0 1-1.498-.075l.275-5.5a.75.75 0 0 1 .786-.711Z"
                fill-rule="evenodd"
              />
            </svg>
            Delete
          </button>
        </li>
      </ul>
    </div>
    <div
      class="flex gap-2"
    >
      <dialog
        class="modal"
        data-testid="modal"
        id=":r0:"
      >
        <div
          class="modal-box"
        >
          <h1
            class="text-xl font-bold"
          >
            Delete Chat
          </h1>
          <p
            class="mt-4"
          >
            Are you sure you want to delete this chat? This action is non-recoverable
          </p>
          <div
            class="modal-action"
          >
            <button
              class="btn"
              data-testid="delete-confirmation-dialog-cancel-btn"
            >
              Cancel
            </button>
            <button
              class="btn btn-error"
              data-testid="delete-confirmation-dialog-delete-btn"
            >
              Delete
            </button>
          </div>
        </div>
      </dialog>
    </div>
  </div>
</div>
`;
